### PR TITLE
Reading nlte_excitation_species from config

### PR DESCRIPTION
**Pull request by sonachitchyan**
### :pencil: Description

**Type:** :beetle: `bugfix` | :rocket: `feature` | :biohazard: `breaking change` | :vertical_traffic_light: `testing` | :memo: `documentation` | :roller_coaster: `infrastructure`

This PR makes sure that `nlte_excitation_species` is read from the config file and passed to the class that constructs the `rate_matrix_index` which will then be used for NLTE excitation treatment.

Also, link issues affected by this pull request by using the keywords: `close`, `closes`, `closed`, `fix`, `fixes`, `fixed`, `resolve`, `resolves` or `resolved`. 

### :pushpin: Resources

Examples, notebooks, and links to useful references.

### :vertical_traffic_light: Testing

How did you test these changes?

- [ ] Testing pipeline
- [ ] Other method (describe)
- [ ] My changes can't be tested (explain why)

### :ballot_box_with_check: Checklist

- [x] I requested two reviewers for this pull request
- [ ] I updated the documentation according to my changes
- [ ] I built the documentation by applying the `build_docs` label

> **Note:** If you are not allowed to perform any of these actions, ping (@) a contributor.

### CODE REVIEW

File: tardis/io/tests/test_config_reader.py [hunk 1]
```diff
@@ -221,6 +221,46 @@ def test_plasma_nlte_section_config(tardis_config_verysimple_nlte):
...
+def test_plasma_nlte_exc_section_config(tardis_config_verysimple_nlte):
+    """
+    Configuration Validation Test for Plasma Section of the Tardis Config YAML File.
+
+    Validates:
+        nlte_excitation_species: should be included in continuum_interaction
+
+    Parameter
+    ---------
+        `tardis_config_verysimple` : YAML File
```
Comment (andrewfullard) [inline tardis/io/tests/test_config_reader.py:233]: _nlte missing

Comment (sonachitchyan) [reply tardis/io/tests/test_config_reader.py]: fixed, thanks for noticing!

**Review by andrewfullard — DISMISSED** [review]
Looks good, tests pass, I'm happy

File: tardis/io/tests/test_config_reader.py [hunk 2]
```diff
@@ -221,6 +221,46 @@ def test_plasma_nlte_section_config(tardis_config_verysimple_nlte):
...
+
+        for species in nlte_excitation_species:
+            if not (
+                species
+                in tardis_config_verysimple_nlte["plasma"][
+                    "continuum_interaction"
+                ]["species"]
+            ):
+                raise ValueError("Nlte excitation species not in continuum.")
+    assert ve.type is ValueError
```
Comment (chvogl) [inline tardis/io/tests/test_config_reader.py:261]: I am very confused about what this test is supposed to do. You set values for the `continuum_interaction_species` and the `nlte_excitation_species` in the config by hand. Then you check that the `nlte_excitation_species` are not in the `continuum_interaction_species`  with a piece of code that only exists in this test. This check fails because the values were set correctly in the config and you raise a `ValueError` by hand. The test, which expects a `ValueError`,  then obviously passes. As far as I can tell, the only thing that is being tested (in a very convoluted way) is that the config values exist and can be set to the desired values. I know that the test if derived from already existing tests but they also do not seem to make much more sense. Maybe I am just missing something?

File: tardis/plasma/standard_plasmas.py [hunk 3]
```diff
@@ -175,15 +180,33 @@ def assemble_plasma(config, model, atom_data=None):
...
-                if not (species in config.plasma.continuum_interaction.species):
-                    raise PlasmaConfigError(
-                        f"NLTE ionization species {species} not in continuum species."
-                    )
+        if (
+            config.plasma.nlte_ionization_species
+            or config.plasma.nlte_excitation_species
+        ):
+            nlte_ionization_species = []
+            nlte_excitation_species = []
```
Comment (chvogl) [inline tardis/plasma/standard_plasmas.py:188]: Why is this needed?

File: tardis/plasma/standard_plasmas.py [hunk 4]
```diff
@@ -175,15 +180,33 @@ def assemble_plasma(config, model, atom_data=None):
...
-            nlte_ionization_species = config.plasma.nlte_ionization_species
-            for species in nlte_ionization_species:
-                if not (species in config.plasma.continuum_interaction.species):
-                    raise PlasmaConfigError(
-                        f"NLTE ionization species {species} not in continuum species."
-                    )
+        if (
+            config.plasma.nlte_ionization_species
+            or config.plasma.nlte_excitation_species
+        ):
```
Comment (chvogl) [inline tardis/plasma/standard_plasmas.py:186]: Why is this if clause needed? You later do individual checks `if config.plasma.nlte_ionization_species` and `if config.plasma.nlte_excitation_species` anyways.

File: tardis/io/tests/test_config_reader.py (hunk 2 above)
Comment (sonachitchyan) [reply tardis/io/tests/test_config_reader.py:261]: I'm not sure why this is the way to do it. I mimicked it from the existing tests. Maybe @andrewfullard would have a good answer?

File: tardis/plasma/standard_plasmas.py (hunk 3 above)
Comment (sonachitchyan) [reply tardis/plasma/standard_plasmas.py:188]: so that if one of them is empty, it passes an empty list.

File: tardis/plasma/standard_plasmas.py (hunk 4 above)
Comment (sonachitchyan) [reply tardis/plasma/standard_plasmas.py:186]: As I am saving them in the same plasma property, I added this part to make sure that that property is added no matter if one of them is empty or not.

File: tardis/plasma/standard_plasmas.py (hunk 3 above)
Comment (chvogl) [reply tardis/plasma/standard_plasmas.py:188]: I see. I think you could also achieve this by moving the assignment `nlte_ionization_species = config.plasma.nlte_ionization_species` and `nlte_excitation_species = config.plasma.nlte_excitation_species` here since the default value for both is an empty list (as defined in the yaml schema).

File: tardis/plasma/standard_plasmas.py (hunk 4 above)
Comment (chvogl) [reply tardis/plasma/standard_plasmas.py:186]: Thanks for clarifying. That makes sense. I guess you could still get rid of the if clause if you do the the assignments `nlte_ionization_species = config.plasma.nlte_ionization_species` and `nlte_excitation_species = config.plasma.nlte_excitation_species` regardless of whether they are actually needed. However, I don't think this a much better solution.

Comment (sonachitchyan) [reply tardis/plasma/standard_plasmas.py:186]: so, I got rid of those 2 lines, and the code still seems to be working.

Comment (sonachitchyan) [reply tardis/plasma/standard_plasmas.py:186]: I got rid of the reassignment for `nlte_ionization_species` and `nlte_excitation_species`. Is there anything else I should change?

Comment (chvogl) [reply tardis/plasma/standard_plasmas.py:186]: I think you need to move the assignments `nlte_ionization_species = config.plasma.nlte_ionization_species` and `nlte_excitation_species = config.plasma.nlte_excitation_species` out of the inner two if clauses. Otherwise, these variables are not defined if they the values from the config are empty lists. Alternatively, you can also use `config.plasma.nlte_excitation_species` and `config.plasma.nlte_ionization_species` in line 206/207.

File: tardis/io/tests/test_config_reader.py (hunk 2 above)
Comment (wkerzendorf) [reply tardis/io/tests/test_config_reader.py:261]: To do this properly we should follow this

def test_fails_but_bad_style():
    try:
        x = 1 / 1
        assert False
    except ZeroDivisionError:
        assert True

**Review by wkerzendorf — DISMISSED** [review]

File: tardis/io/tests/test_config_reader.py (hunk 2 above)
Comment (sonachitchyan) [reply tardis/io/tests/test_config_reader.py]: I updated the tests, now they get skipped, and I'm not sure if that's a good or a bad thing. I pushed it here, so that you guys can see them too.

File: tardis/io/tests/test_config_reader.py [hunk 5]
```diff
@@ -190,35 +196,55 @@ def test_plasma_nlte_section_config(tardis_config_verysimple_nlte):
...
+        "species"
+    ] = [
+        "He I",
+    ]
+    tardis_config_verysimple_nlte["plasma"]["nlte_excitation_species"] = ["H I"]
+    with pytest.raises(PlasmaConfigError) as ve:
+        plasma = assemble_plasma(
+            tardis_config_verysimple_nlte, nlte_raw_model, nlte_atom_data
+        )
+    assert ve.type is PlasmaConfigError
```
Comment (chvogl) [inline tardis/io/tests/test_config_reader.py:247]:
```suggestion
```
The type of the Exception is already checked by the `pytest.raises`. I don't think checking the type after the with clause adds anything but I might be wrong.

File: tardis/io/tests/test_config_reader.py [hunk 6]
```diff
@@ -190,35 +196,55 @@ def test_plasma_nlte_section_config(tardis_config_verysimple_nlte):
...
-                    "continuum_interaction"
-                ]["species"]
-            ):
-                raise ValueError("Nlte ionization species not in continuum.")
-    assert ve.type is ValueError
+    with pytest.raises(PlasmaConfigError) as ve:
+        plasma = assemble_plasma(
+            tardis_config_verysimple_nlte, nlte_raw_model, nlte_atom_data
+        )
+    assert ve.type is PlasmaConfigError
```
Comment (chvogl) [inline tardis/io/tests/test_config_reader.py:217]:
```suggestion
```
The type of the Exception is already checked by the pytest.raises. I don't think checking the type after the with clause adds anything but I might be wrong.

File: tardis/io/tests/test_config_reader.py [hunk 7]
```diff
@@ -190,35 +196,55 @@ def test_plasma_nlte_section_config(tardis_config_verysimple_nlte):
...
-                in tardis_config_verysimple_nlte["plasma"][
-                    "continuum_interaction"
-                ]["species"]
-            ):
-                raise ValueError("Nlte ionization species not in continuum.")
-    assert ve.type is ValueError
+    with pytest.raises(PlasmaConfigError) as ve:
+        plasma = assemble_plasma(
+            tardis_config_verysimple_nlte, nlte_raw_model, nlte_atom_data
+        )
```
Comment (chvogl) [inline tardis/io/tests/test_config_reader.py:216]: The new tests make much more sense than the old ones but aren't they strictly speaking plasma tests?
What do other people think?

File: tardis/io/tests/test_config_reader.py [hunk 8]
```diff
@@ -190,35 +196,55 @@ def test_plasma_nlte_section_config(tardis_config_verysimple_nlte):
...
-                species
-                in tardis_config_verysimple_nlte["plasma"][
-                    "continuum_interaction"
-                ]["species"]
-            ):
-                raise ValueError("Nlte ionization species not in continuum.")
-    assert ve.type is ValueError
+    with pytest.raises(PlasmaConfigError) as ve:
+        plasma = assemble_plasma(
+            tardis_config_verysimple_nlte, nlte_raw_model, nlte_atom_data
```
Comment (chvogl) [inline tardis/io/tests/test_config_reader.py:215]: I think you need to pass a Configuration object not a config_dict?

Comment (sonachitchyan) [reply tardis/io/tests/test_config_reader.py:215]: you were right! I'll push the fixed version once the tests pass on my local machine.

**Review by andrewfullard — APPROVED** [review]

File: tardis/io/tests/test_config_reader.py [hunk 9]
```diff
@@ -190,35 +196,55 @@ def test_plasma_nlte_section_config(tardis_config_verysimple_nlte):
...
+):
+    """
+    Configuration Validation Test for Plasma Section of the Tardis Config YAML File.
+
+    Validates:
+        nlte_excitation_species: should be included in continuum_interaction
+
+    Parameter
+    ---------
+        `tardis_config_verysimple_nlte` : YAML File
```
[the remark this answers is not in the document]
Comment (chvogl) [reply tardis/io/tests/test_config_reader.py:229]:
```suggestion
        `tardis_config_verysimple_nlte` : YAML File
        `nlte_raw_model` : A simple model
        `nlte_atom_data` : An example atomic dataset
```

File: tardis/io/tests/test_config_reader.py (hunk 7 above)
Comment (chvogl) [reply tardis/io/tests/test_config_reader.py]: Nobody seems to care. Let's move on.

**Review by chvogl — APPROVED** [review]

**Comment by sonachitchyan** [comment]
So the tests failed out of the blue. I will look into it.

**Comment by chvogl** [comment]
> So the tests failed out of the blue. I will look into it.

This looks extremely similar to #2191.

**Comment by sonachitchyan** [comment]
> > So the tests failed out of the blue. I will look into it.
> 
> This looks extremely similar to #2191.

yes, the error happens at the exact same spot. @atharva-2001 do you know why this might be happening?

**Comment by chvogl** [comment]
This might not be particularly useful but one difference between the failed and successful test runs is that for the failed runs the cached conda environment is not found. This then triggers the "Update environment" step.

**Comment by sonachitchyan** [comment]
> This might not be particularly useful but one difference between the failed and successful test runs is that for the failed runs the cached conda environment is not found. This then triggers the "Update environment" step.

I made a separate PR, to try to see if I can fix this. If that works, it should all be fine afterwards.

**Review by andrewfullard — APPROVED** [review]
